python/lsst/resources/s3.py

```diff
@@ -114,6 +127,28 @@ def __call__(self, bytes_amount: int) -> None:
             )
 
 
+def _translate_client_error(err: ClientError, fallback: Exception) -> None:
+    """Translate a ClientError into a specialist error else raise.
+
+    Parameters
+    ----------
+    err : `ClientError`
+        Exception to translate.
+    fallback : `Exception`, optional
+        Exception to raise if not intercepted. If `None` the original
+        exception will be re-raised.
+    """
+    if "(429)" in str(err):
+        # ClientError includes the error code in the message
+        # but no direct way to access it without looking inside the
+        # response.
+        raise _TooManyRequestsException(str(err)) from err
+    elif fallback:
```

Comment:
I'm not sure I like this.  Can't some ClientErrors be due to problems that are not caused by a missing object?

Reply:
Yes. I'm not sure how to deal with that without putting explicit handlers in the three places. I don't know the breadth of what can end up in the ClientError so I'm not sure what the possibilities are.

Reply:
Why have the fallback at all?  Just return the original ClientError if it's not translatable.

Reply:
The caller isn't really expecting a special S3 exception (and might not even know S3 is involved) and we mostly try to translate things. The `from err` will ensure that the previous exception is visible but given you think it's preferred I'll just put in a `raise` in the `except` block and not try to do anything special.